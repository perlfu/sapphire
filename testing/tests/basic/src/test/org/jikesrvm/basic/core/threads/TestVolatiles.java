/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Common Public License (CPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/cpl1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package test.org.jikesrvm.basic.core.threads;

class TestVolatiles extends XThread {

  public static void main(String[] args) {
    for (int i = 0; i < 5; i++) {
      TestVolatiles tv = new TestVolatiles(i);
      tv.start();
    }
    XThread.say("bye");
    XThread.outputMessages();
  }

  static volatile long vl = 0;
  static volatile int vi = 0;

  int n;
  long l;

  TestVolatiles(int i) {
    super("V" + i);
    n = i;
    l = (((long) n) << 32) + n;
  }

  void performTask() {
    int errors = 0;
    for (int i = 0; i < 10000000; i++) {
      long tl = vl;
      vl = l;
      int n0 = (int) tl;
      int n1 = (int) (tl >> 32);
      if (n0 != n1) errors++;
      vi = n;
    }
    tsay(errors + " errors found");
  }
}
