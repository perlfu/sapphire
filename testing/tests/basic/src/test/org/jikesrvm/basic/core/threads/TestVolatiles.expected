Thread 0 Main: bye
Thread 1 V0: creating
Thread 1 V0: starting
Thread 1 V0: run starting
Thread 1 V0: 0 errors found
Thread 1 V0: run finishing
Thread 2 V1: creating
Thread 2 V1: starting
Thread 2 V1: run starting
Thread 2 V1: 0 errors found
Thread 2 V1: run finishing
Thread 3 V2: creating
Thread 3 V2: starting
Thread 3 V2: run starting
Thread 3 V2: 0 errors found
Thread 3 V2: run finishing
Thread 4 V3: creating
Thread 4 V3: starting
Thread 4 V3: run starting
Thread 4 V3: 0 errors found
Thread 4 V3: run finishing
Thread 5 V4: creating
Thread 5 V4: starting
Thread 5 V4: run starting
Thread 5 V4: 0 errors found
Thread 5 V4: run finishing
